openshift_metrics/utils.py

```diff
@@ -0,0 +1,144 @@
+#   Licensed under the Apache License, Version 2.0 (the "License"); you may
+#   not use this file except in compliance with the License. You may obtain
+#   a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#   WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#   License for the specific language governing permissions and limitations
+#   under the License.
+#
+
+import datetime
+import json
+import requests
+import time
+
+import openshift as oc
+
+
+def query_metric(openshift_url, token, metric, report_date, retry=3):
+    attempt = 0
+    data = None
+    headers = {'Authorization': "Bearer %s" % token}
+    day_url_vars = "start=%sT00:00:00Z&end=%sT23:59:59Z" % (report_date, report_date)
+    print("Retrieving metric: %s" % metric)
+    while attempt < retry:
+        url = "%s/api/v1/query_range?query=%s&%s&step=60s" % (openshift_url, metric, day_url_vars)
+        r = requests.get(url, headers=headers, verify=False)
+        if r.status_code != 200:
+            print("%s Response: %s" % (r.status_code, r.reason))
+        else:
+            data = r.json()['data']['result']
+            if data:
+                break
+            else:
+                print("Empty result set")
+        attempt = attempt + 1
+        time.sleep(3)
+    if not data:
+        raise Exception('Error retrieving metric: %s' % metric)
+    return data
+
+def get_namespace_annotations():
+    namespaces_dict = {}
+    namespaces = oc.selector("namespaces").objects()
+    for namespace in namespaces:
+        namespace_dict = json.loads(namespace.as_json())['metadata']
+        namespaces_dict[namespace_dict['name']] = namespace_dict['annotations']
+    return namespaces_dict
+
+def merge_metrics(metric_name, metric_list, output_dict):
+    for metric in metric_list:
+        pod = metric['metric']['pod']
+        if not output_dict.has_key(pod):
+            output_dict[pod] = {'namespace': metric['metric']['namespace'],
+                                'metrics': {}}
+        for value in metric['values']:
+            epoch_time = value[0]
+            if not output_dict[pod]['metrics'].has_key(epoch_time):
+                output_dict[pod]['metrics'][epoch_time] = {}
+            output_dict[pod]['metrics'][epoch_time][metric_name] = value[1]
+    return output_dict
+
+def condense_metrics(input_metrics_dict, metrics_to_check):
+    condensed_dict = {}
+    for pod in input_metrics_dict:
+        pod_dict = input_metrics_dict[pod]
+        metrics_dict = pod_dict['metrics']
+        new_metrics_dict = {}
+        epoch_times_list = metrics_dict.keys()
+        epoch_times_list.sort()
```

Comment:
What version of Python are you working with? In 3.8 and later (and possibly earlier, I got bored after trying 3.10, 3.9, and 3.8), the `keys()` method returns a `dict_keys` object rather than a list, and this fails with:

```
AttributeError: 'dict_keys' object has no attribute 'sort'
```

You would need to write instead:

```
        epoch_times_list = list(metrics_dict.keys())
        epoch_times_list.sort()
```

Or better yet, don't use the in-place sort method and instead use the `sorted` built-in:

```
epoch_times_list = sorted(metrics_dict.keys())
```

Reply:
Whoooops I was indeed using python 2. Fixed and updated!